TestResultSummaryService/perf/DataManagerAggregate.js

```diff
@@ -4,68 +4,112 @@ const ObjectID = require( 'mongodb' ).ObjectID;
-            name = childBuild.tests[0].benchmarkName;
-            variant = childBuild.tests[0].benchmarkVariant;
-            for ( let {testData} of childBuild.tests){
-                if ( Array.isArray(testData.metrics) ) {
-                    for ( let {name, value} of testData.metrics ){
-                        benchmarkMetricsCollection[name] = benchmarkMetricsCollection[name] || [];
-                        if ( Array.isArray(value) ) {
-                            benchmarkMetricsCollection[name] = benchmarkMetricsCollection[name].concat(value);
+            for ( let {benchmarkName, benchmarkVariant, testData} of childBuild.tests){
+                if (benchmarkName && benchmarkVariant && testData && Array.isArray(testData.metrics) && testData.metrics.length > 0){
+                    // Example: Jenkins build (running 2 benchmarks with 2 iterations) in this order: Benchmark_1, Benchmark_2, Benchmark_1, Benchmark_2.
+                    let name_variant_keysArray = Object.keys(aggRawMetricValues);
+                    let name_varirant_key = benchmarkName+"&&"+benchmarkVariant;//use benchmarkName and benchmarkVariant to make a unique name_variant key
+                    /**
+                     * Check aggRawMetricValues JSON object and see if it could include an existing name_variant key,
+                     * If yes, we should concatenate our raw value collections to the end of the existing name_variant value
+                     * Otherwise, we should create a new object.
+                     */
+                    if (name_variant_keysArray && name_variant_keysArray.length > 0 && name_variant_keysArray.includes(name_varirant_key)) {
+                        for ( let {name, value} of testData.metrics ){
+                            aggRawMetricValues[name_varirant_key][name] = aggRawMetricValues[name_varirant_key][name] || [];
+                            if ( Array.isArray(value) ) {
+                                aggRawMetricValues[name_varirant_key][name] = aggRawMetricValues[name_varirant_key][name].concat(value);
+                            }
+                            aggRawMetricValues[name_varirant_key][name] = aggRawMetricValues[name_varirant_key][name].filter (n => n);
+                        }
+                    } else {
+                        aggRawMetricValues[name_varirant_key] = {};
+                        for ( let {name, value} of testData.metrics ){
+                            aggRawMetricValues[name_varirant_key][name] = aggRawMetricValues[name_varirant_key][name] || [];
+                            if ( Array.isArray(value) ) {
+                                aggRawMetricValues[name_varirant_key][name] = aggRawMetricValues[name_varirant_key][name].concat(value);
+                            }
+                            aggRawMetricValues[name_varirant_key][name] = aggRawMetricValues[name_varirant_key][name].filter (n => n);
```

Comment:
Except for one line under the else statement, all other lines are exactly the same as those under if statement. It's redundant! We should do this:
If key missing, add it.
For all (newly added or the ones from before), aggregate.

Reply:
Agree!

Reply:
This is still an issue. Can we please fix this?

Reply:
@sophiaxu0424 I still see some redundancy there!